File: java/189.java
class Solution {
    public void rotate(int[] nums, int k) {
        int l = nums.length;
        k = k % l;
        reverse(nums,0,l-k-1);
        reverse(nums,l-k,l-1);
        reverse(nums,0,l-1);
        
    }
    
    public void reverse(int[] nums,int l,int r){
        while(l < r){
            int temp = nums[l];
            nums[l] = nums[r];
            nums[r] = temp;
            l ++;
            r --;
        }
    }
}

File: go/189.go
func rotate(nums []int, k int) {
	l := len(nums)
	k = k % l
	swap(nums, 0, l-1)
	swap(nums, 0, k-1)
	swap(nums, k, l-1)
}

func swap(nums []int, i int, j int) {
	for i <= j {
		tmp := nums[i]
		nums[i] = nums[j]
		nums[j] = tmp
		i++
		j--
	}
}